Built font-awesome icon picker

ewels opened the pull request:
My version of #333, inspired by @mashehu 

Differences to #333:

* Homebrew - no additional dependencies
* Fully functional (updates schema properly)
* Support for keyboard shortcuts / behaviour you might expect (click to hide, escape to hide etc)
* Focus support - auto focus of search field, refocus of button when popover hidden

I know you like the caret @mashehu, but I'm mostly working on my small laptop screen where space is at a massive premium and I preferred it without.

Additional changes:

* Form is now full-width fluid, instead of constraining to the page container
* Stripped font-awesome icon from modal
* JSON Schema now always shown at the bottom, instead of having to click _Finished_ to see it.

Note that I used your python script to generate the JSON file @mashehu, with a couple of changes. Firstly to give a custom JSON file as I'm not using iconpicker, and also simplifying the output a bit - I managed to reduce the file size by about half by making it a simple key:value map instead of more complex arrays etc.

Demo using only keyboard-shortcuts:

![icon_picker](https://user-images.githubusercontent.com/465550/77053095-18e40e80-69ce-11ea-9bd5-a8aa5bcf83cc.gif)

ewels commented:
@mashehu - if we're happy with this implementation, I suggest merging it and then refactoring your PR. I didn't want to commit your script and take credit for it 😅 If we keep your bits like moving the imports around, the script etc. Then once that's merged I'll make a PR against your script with the changes that I made. Sound ok?

mashehu commented on public_html/assets/css/nf-core.css:
```diff
@@ -906,6 +923,25 @@ Based on https://codepen.io/wintr/pen/beBJBb */
 }
 
+.fa_icon_picker.popover {
+  width: 315px;
```
why in `px`? don't the icons scale with the rest of the fonts, so use `rem` instead?

ewels replied:
Yeah, probably should do. Got to the value by tweaking in the browser so was being lazy. I'll fix 👍

ewels replied:
Ok, switched to `rem` 👍

mashehu commented on public_html/assets/js/fa-icons.json:
```diff
@@ -0,0 +1 @@
+{"fab fa-500px": "", "fab fa-accessible-icon": "accessibility handicap person wheelchair wheelchair-alt", "fab fa-accusoft": "", "fab fa-acquisitions-incorporated": "Dungeons & Dragons d&d dnd fantasy game gaming tabletop", "fas fa-ad": "advertisement media newspaper promotion publicity", "fas fa-address-book": "contact directory index little black book rolodex", "far fa-address-book": "contact d ...
```
for entries without a search term (e.g. 500px), the value should be the second part of the key without the `fa-`. Otherwise they will not come up in the search, I think.

mashehu replied:
Actually, we should add that, to all entries, otherwise you will not find exact matches on the name,. e.g. "x-ray" (if I understand the search function correctly)

ewels replied:
Not needed, as I made the search box look both in the search terms but also the class names: `.popover-body div button[data-classname*='"+search_term+"']`

https://github.com/nf-core/nf-co.re/pull/334/files#diff-50fe998361a51c27bb42f5df30467eeaR90

mashehu commented on public_html/assets/js/nf-core-schema-builder.js:
```diff
@@ -469,20 +576,6 @@ $(function () {
                 e.stopPropagation();
             }
         }
-        // Check that the font-awesome icon looks right
```
will the schema validation do that, when somebody uploads their schema with a malformed FA icon?

ewels replied:
Currently there is no validation so if someone uploads something weird then it will, yeah, probably break.

ewels replied:
Ok, I added a validation step when building the schema rows. Now if there's something weird it will just delete the key from the schema and treat it as if there was no icon. It throws an error message to the browser console.

mashehu commented on public_html/assets/js/nf-core-schema-builder.js:
```diff
@@ -695,22 +788,23 @@ function generate_param_row(id, param){
         </div>
-        <div class="col-sm-auto align-self-center schema_row_help_text_icon">`+help_text_icon+`</div>
-        <div class="col-sm-1">
+        <div class="col-auto align-self-center schema_row_help_text_icon iconpicker-component">`+help_text_icon+`</div>
```
Why is this class needed here?

ewels replied:
Yeah, it wasn't! I noticed it as well, now removed. 👍

ewels commented:
Ok, I added a new modal for the help text. So the settings modal is trimmed down a lot more now which feels better.

Help modal follows the same kind of UI as the icon picker in terms of keyboard shortcuts and focus niceties. I also added GitHub-style _Write_ and _Preview_ tabs so that it's easy to see how it will be rendered:

![helptext_modal](https://user-images.githubusercontent.com/465550/77065377-7766b800-69e1-11ea-899f-1c23a72afced.gif)

mashehu commented on public_html/assets/css/nf-core.css:
```diff
@@ -906,6 +920,29 @@ Based on https://codepen.io/wintr/pen/beBJBb */
+  padding: 0;
+}
+.fa_icon_picker .popover-body {
+  height: 200px;
```
found another one 🙂 (I can't test these changes locally and I am bad with guessing, so you have to fix that on your side)

mashehu commented on public_html/json_schema_build.php:
```diff
@@ -249,15 +256,7 @@ function validate_cache_id($cache_id){
-                            some may render as plain text.
-                        </small>
-                    </div>
+                    <p class="text-muted settings_nothing_special">No special settings available.</p>
```
cool classname 😎 😉

mashehu commented on public_html/json_schema_build.php:
```diff
@@ -322,10 +313,67 @@ function validate_cache_id($cache_id){
+                                <pre><span class="helptext-preview-title"></span>
+<span class="helptext-preview-description"></span>
+
+<span class="helptext-preview-helptext text-muted"></span></pre>
```
how about if we call this class `helptext-preview-text`, sound less confusing to me 🙂

ewels replied:
Yeah, could do. It makes more sense in the javascript though as the variable we put in here is called `helptext`
